test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1

```diff
@@ -1098,6 +1098,25 @@ class InheritedClassTest : System.Attribute
             $expectedPath = Join-Path $PSScriptRoot -ChildPath BugFix.Tests.ps1
             $res.CompletionMatches[0].CompletionText | Should -Be "`"$expectedPath`""
         }
+
+        It "Should keep '~' in completiontext when it's used to refer to home in input" {
```

Comment:
Seems like there should be tests covering `EscapeCharIfNeeded()`.  I think you can easily do this as TestCases, pseudo-code:

```powershell
It 'Special PowerShell characters in the path are handled correctly: <char>' -TestCases @(
  @{ char = '$' }
  @{ char = '@' }
  @{ char = '`u' }
  # the rest of them here
) {
  param($char)
  
  $tempDirName = "${char}test${char}"
  $tempDirPath = Join-Path $testdrive $tempDirName
  $tempDir = New-Item -Path $tempDirPath -ItemType Directory -Force
  $res = TabExpansion2 -inputScript $char
  $res.CompletionMatches[0].CompletionText | Should -BeExactly ".${seperator}${tempDirName}"
}

Reply:
While writing the test I realized I had made a mistake and don't actually need to explicitly handle the special PS characters because I should be escaping the backtick anyway. I've updated the code and added tests for the new logic, please take another look at it.